mlflow/fastai/callback.py

```diff
@@ -0,0 +1,157 @@
+import numpy as np
+import os
+import shutil
+import tempfile
+from functools import partial
+import matplotlib.pyplot as plt
+
+import mlflow.tracking
+from mlflow.utils.autologging_utils import try_mlflow_log
+from mlflow.fastai import log_model
+
+from fastai.callback.core import Callback
+
+
+# Move outside, because it cannot be pickled. Besides, ExceptionSafeClass was giving some issues
+class __MLflowFastaiCallback(Callback):
+    """
+    Callback for auto-logging metrics and parameters.
+    Records model structural information as params when training begins.
+    """
+
+    from fastai.learner import Recorder
+    from fastai.callback.all import TrackerCallback
+
+    remove_on_fetch, run_before, run_after = True, TrackerCallback, Recorder
+
+    def __init__(self, metrics_logger, log_models, is_fine_tune=False):
+        super().__init__()
+        self.metrics_logger = metrics_logger
+        self.log_models = log_models
+        self.is_fine_tune = is_fine_tune
+        self.freeze_prefix = ""
+
+    def after_epoch(self):
+        """Log loss and other metrics values after each epoch"""
+
+        def _is_float(x):
+            try:
+                float(x)
+                return True
+            except (ValueError, TypeError):
+                return False
+
+        # Do not record in case of predicting
+        if hasattr(self, "lr_finder") or hasattr(self, "gather_preds"):
+            return
+        # Remove non-float metrics and record the rest.
+        metrics = self.recorder.log
+        metrics = dict(zip(self.recorder.metric_names, metrics))
+        for key in [key for key, val in metrics.items() if not _is_float(val)]:
+            del metrics[key]
+        self.metrics_logger.record_metrics(metrics, step=metrics["epoch"])
+
+    def before_fit(self):
+        from fastai.callback.all import ParamScheduler
+
+        # Do not record in case of predicting or lr_finder
+        if hasattr(self, "lr_finder") or hasattr(self, "gather_preds"):
+            return
+
+        if self.is_fine_tune and len(self.opt.param_lists) == 1:
+            print(
+                """ WARNING: Using `fine_tune` with model which cannot be frozen.
+                Current model have only one param group which makes it impossible to freeze.
+                Because of this it will record some fitting params twice (overriding exception)
+                """
+            )
+
+        frozen = self.opt.frozen_idx != 0
+        if frozen and self.is_fine_tune:
+            self.freeze_prefix = "freeze_"
+        else:
+            self.freeze_prefix = ""
```

Comment:
Can we just log `frozen_idx` (that indicates the model was trained with some layers frozen) rather than prefixing all logged parameters with `freeze_`?

Reply:
It looks like both "freeze_{param}" and "{param}" are to be logged when fine_tune is used. I'm not familiar with fastai. Unsure if it's useful that way. You can see this behavior tested in tests/fastai/test_fastai_autolog.py.

Reply:
Here's the source code of `fine_tune`:

```python
@patch
@delegates(Learner.fit_one_cycle)
def fine_tune(self:Learner, epochs, base_lr=2e-3, freeze_epochs=1, lr_mult=100,
              pct_start=0.3, div=5.0, **kwargs):
    "Fine tune with `Learner.freeze` for `freeze_epochs`, then with `Learner.unfreeze` for `epochs`, using discriminative LR."
    self.freeze()
    self.fit_one_cycle(freeze_epochs, slice(base_lr), pct_start=0.99, **kwargs)
    base_lr /= 2
    self.unfreeze()
    self.fit_one_cycle(epochs, slice(base_lr/lr_mult, base_lr), pct_start=pct_start, div=div, **kwargs)
```

https://github.com/fastai/fastai/blob/4b0785254fdece1a44859956b6e54eedb167a97e/fastai/callback/schedule.py#L155-L164

Reply:
Logged params look like this:

<img width="50%" src="https://user-images.githubusercontent.com/17039389/133068969-8a07d19a-27ce-4506-ae53-911dedf9ea4d.png" />

Reply:
A plot for `freeze_lr` looks like this:

<img width="50%" src="https://user-images.githubusercontent.com/17039389/133072511-ce4af7d2-568c-496b-a1d5-aae44cbf5cb7.png" />